### PR TITLE
Add support for passive event listeners according to google

### PR DESCRIPTION
**Pull request by nicolaskopp**
https://developers.google.com/web/tools/lighthouse/audits/passive-event-listeners

working fiddle:
http://jsfiddle.net/qnsdy5d7/

### CODE REVIEW

**Comment by KalleVuorjoki** [comment]
Thanks for the PR! I tested this PR with with mobile (Brave + Chrome) and unfortunately the swiping did not work after this patch.

**Comment by nicolaskopp** [comment]
Bummer, an weird that it does not work on mobile. I'll look into it.

**Comment by djkrishnasty** [comment]
@nicolaskopp Thank you for working on this. I think the problem is the `addEventListener` API doesn't support event namespaces such as `touchstart.slick` so the touch related events never fire. To fix this you will have to add event listeners on just the events such as `el.addEventListener('touchstart', ...)` but I am not sure if we want to remove namespacing.

**Comment by djkrishnasty** [comment]
@nicolaskopp Also you need to invoke methods `_.isTouch` and `_.canListenPassive` to get their values but I don't think we need to check if it's a touch device because the passive option should be ignored on non-touch devices anyway.

**Comment by Anikethana** [comment]
any updates on this ?

**Comment by Quazz** [comment]
Adding

`	jQuery.event.special.touchmove = {
  setup: function( _, ns, handle ){
    if ( ns.includes("noPreventDefault") ) {
      this.addEventListener("touchmove", handle, { passive: false });
    } else {
      this.addEventListener("touchmove", handle, { passive: true });
    }
  }
};`

before `Slick.prototype.initializeEvents = function() {` works for me

**Comment by sowinski** [comment]
Please merge and create a new release :)

**Comment by Chris9046** [comment]
@nicolaskopp Thanks for the PR! Do you have any updates on the issue with swipe

**Comment by nicolaskopp** [comment]
Hi @Chris9046,

I'm still alive! Yes, actually I'm going to look into it today and hopefully push a working MR which sorts out the swiping issue.

Stay tuned!

**Comment by nicolaskopp** [comment]
New working fiddle:

http://jsfiddle.net/x8s4hztn/

Thanks to @krishna-veerareddy  and @Quazz  for pointing me into the right direction.

I opted neither to override jQuery behavior nor to purge namespaces, but instead to override EventTarget handling int he JS core.

See     https://stackoverflow.com/questions/46094912/added-non-passive-event-listener-to-a-scroll-blocking-touchstart-event?rq=1
   
for more information.

MR updated.

**Comment by nicolaskopp** [comment]
@skigo strange, for me the message disappeared completely. Could you check out https://github.com/nicolaskopp/slick and fire up the index.html of that repo and run google audits from within chrome and post results?

Here are my results with my patch:

https://web.tresorit.com/l/PnX4a#42swVNdicYM1T5PmQIQYCQ

and here without the patch:

https://web.tresorit.com/l/hCfed#M9TwjeiIPVljZ7X7U8UvTQ

you can see that without the patch, I have the "Does not use passive listeners to improve scrolling performance" whereas in version 1 (with patch) that warning does not fire up.

Steps to reproduce:

1.) `git clone https://github.com/nicolaskopp/slick`
2.) put repo into some local webserver, or open "index.html" with intelliJ, so that it opens up in a temporary localhost environment
3.) Run "Audits" in google chrome developer tools

Thanks!

**Comment by skigo** [comment]
Hello,
I used your slick.js file and still have the same issue. If I disable slick totally the issue is gone.
Also, i got these error(check the attachment)

________________________________

**Comment by nicolaskopp** [comment]
@skigo there is not attachment that I can see, and to be honest, I'm not really sure what to do now. As I recorded on video, the passive listener issue is gone for me and everything else is working, so I don't really know what to do or test or actually do _differently_, as there is no bug for me to debug.

It would be cool if we could get more people to test this and verify if the issue is indeed resolved or if there are still problems.

Sorry that I cannot be of more help here. Maybe you can try to test the fix as suggested by https://github.com/kenwheeler/slick/pull/3422#issuecomment-436267071 ? I did not try, but I hesitate to override jQuery functionality, because we have to take care of quite a bunch of edge-cases as well as backwards- and forward compatibility with different jQuery versions.

**Comment by skigo** [comment]
Hello mate, i tried everything, i also tried to another project but same issue.

Consider marking your touch and wheel event listeners as `passive` to improve your page's scroll performance. Learn more<https://web.dev/uses-passive-event-listeners?utm_source=lighthouse&utm_medium=devtools>.
 Show 3rd-party resources (0)
URL
Location

webpack:///./assets/js/vendor/slick.js?

line: 49


________________________________

**Comment by nicolaskopp** [comment]
@skigo do you have a website where I can look into with your usecase? Do you have the same issue on the sample project provided by slick?

**Comment by skigo** [comment]
Yea sure, can you give me your email or send me an email at skigo15@hotmail.com so i can provide you the url.

Regards,
D

________________________________

**Comment by nicolaskopp** [comment]
@skigo  mail sent

**Comment by skigo** [comment]
> @skigo mail sent

the link in on your email

**Comment by nicolaskopp** [comment]
@skigo I did not receive an email from you. I sent you a mail yesterday. You can also issue-3422@niolaskopp.de to send me the link and I will take a look.

**Comment by Chris9046** [comment]
@nicolaskopp  Thank you for working on this. I tested this but I had the same issue. I tried the fix as suggested by @Quazz and overwrite also the touchstart and worked for me.

` jQuery.event.special.touchstart = {
		setup: function (_, ns, handle) {
			if (ns.includes('noPreventDefault')) {
				this.addEventListener('touchstart', handle, { passive: false });
			} else {
				this.addEventListener('touchstart', handle, { passive: true });
			}
		}
	}; `


`jQuery.event.special.touchmove = {
		setup: function (_, ns, handle) {
			if (ns.includes('noPreventDefault')) {
				this.addEventListener('touchmove', handle, { passive: false });
			} else {
				this.addEventListener('touchmove', handle, { passive: true });
			}
		}
	};`

**Comment by nicolaskopp** [comment]
Thank you @Chris9046 .

Although I did not have the issue anymore, I made a new commit, which integrates your suggested changes.

New working fiddle:

http://jsfiddle.net/5r2gc3vp/

also integrated into https://github.com/nicolaskopp/slick .

@skigo does it work for you now?

**Comment by naton** [comment]
FWIW, the passive listener warning is gone for me now, increasing my BP score by +7 :)

**Comment by nicolaskopp** [comment]
@naton thanks for the verification! Anyway, as @kenwheeler  has not been active in this repo for 11 months or so, I put up a NPM installation instruction on https://github.com/nicolaskopp/slick .

I will make sure to update my repo with the main trunk regularly, until this MR gets merged.

**However, I would advise that everyone should use the kenwheeler/slick for production if you don't need passive listener support.**

**Comment by SohanChotia** [comment]
Hi @nicolaskopp , the use of passive listener is mostly for PageSpeed thing and would boost the score too. 

So is your version safe to use for the live website or not, as you gave a warning/advise ...

**Comment by nicolaskopp** [comment]
Hi @SohanChotia,

it is absolutely safe to use my fork [at the time of me writing this], however, I cannot guarantee that I will be able to keep my fork up to date with all bug- and security-fixes of the original branch in the future. I will check the main branch regularly and merge changes, but it is not my focus to be a maintainer.

This is why I suggested to use the original branch as often as possible if you are not dependent on pagespeed.

**Comment by sowinski** [comment]
When will this fix be merged into the origin version?

**Comment by nicolaskopp** [comment]
@sowinski  you have to ask @kenwheeler , it's integrated on my fork: https://github.com/nicolaskopp/slick .

**Comment by doughballs** [comment]
This is not working for me - used npm install, if I reference slick.min.js the sliders works as usual but the violation error still flags, if I use slick.js then the slider breaks and I get a  `Uncaught TypeError: $(...).slick is not a function`.

**Comment by nicolaskopp** [comment]
> This is not working for me - used npm install, if I reference slick.min.js the sliders works as usual but the violation error still flags, if I use slick.js then the slider breaks and I get a `Uncaught TypeError: $(...).slick is not a function`.
> 
> Why does the minified version not include the fix? And if it should - why are the violations still being logged?

I think this is a bug with my repo. I will raise an issue and fix it there. hold on. Thanks for the heads up!

**Comment by doughballs** [comment]
@nicolaskopp it may not actually be a bug, after some further tests, let me explain:

I compile all my JS into a single file, Slick included - I've always worked this way. When I included your repo over the oroginal, the compiled file was eithe r not working or throwing the errors. If I link to it traditionally, standalone js file, it all works fine. So seems to be something in my build process in grunt. I wouldn't want you to waste any time on what could be an edge case :)

**Comment by nicolaskopp** [comment]
> @nicolaskopp it may not actually be a bug, after some further tests, let me explain:
> 
> I compile all my JS into a single file, Slick included - I've always worked this way. When I included your repo over the original, the compiled file was either not working (still getting violation notice) or throwing the jQuery errors. If I link to it traditionally - the standalone js file - it all works fine. So seems to be something in my build process in grunt. I wouldn't want you to waste any time on what could be an edge case :)

thank you for the follow-up! For your case, these might be some compiler settings which are too aggressive in optimizing? However, I think it's an easy fix to just compile "my" js file and provide a minified (working) js in the repo. That's like 3 seconds for work. The time consuming part is testing it in various resolutions and browsers...Need to find a time slot for that, but I will do it. :)

**Comment by Aulig** [comment]
Very nice, got a nice pagespeed boost from switching. Thanks!

For some reason it seems to add `style="width: 100%; display: inline-block;"` to my carousel slides though. So for example I'd end up with such a carousel, even though I didn't set any style on the slide:
```html
<div class="addon-carousel slick-initialized slick-slider slick-dotted">
    <button class="slick-prev slick-arrow" aria-label="Previous" type="button" style="">Previous</button>
    <div class="slick-list draggable">
        <div class="slick-track" style="opacity: 1; width: 4872px; transform: translate3d(-1392px, 0px, 0px);">
            <div class="slick-slide slick-cloned" data-slick-index="-4" id="" aria-hidden="true" style="width: 348px;" tabindex="-1">
                <div>
                                                           vvvvvvvvvvvvvv THE ISSUE vvvvvvvvvvvvv
                    <div class="card lifting-card mx-3" style="width: 100%; display: inline-block;">
                        <img decoding="async" class="card-img-top p-3 img-fluid" src="..." alt="..." />
                        <div class="card-body">
                            <h3 class="h5 card-title">Title</h3>
                            <p class="card-text">
                                Text
                            </p>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <button class="slick-next slick-arrow" aria-label="Next" type="button" style="">Next</button>
    <ul class="slick-dots" style="" role="tablist">
        <li class="slick-active" role="presentation"><button type="button" role="tab" id="slick-slide-control10" aria-controls="slick-slide10" aria-label="1 of 2" tabindex="0" aria-selected="true">1</button></li>
        <li role="presentation" class=""><button type="button" role="tab" id="slick-slide-control11" aria-controls="slick-slide14" aria-label="2 of 2" tabindex="-1">2</button></li>
    </ul>
</div>
```

Any ideas on what the issue could be? I'm generating the slider with:

```javascript
$(".addon-carousel").slick({
    dots: true,
    infinite: true,
    autoplay: true,
    autoplaySpeed: 20000,
    slidesToShow: 4,
    slidesToScroll: 4,
    responsive: [
        { breakpoint: 1200, settings: { slidesToShow: 3, slidesToScroll: 3 } },
        { breakpoint: 768, settings: { slidesToShow: 2, slidesToScroll: 2 } },
        { breakpoint: 576, settings: { slidesToShow: 1, slidesToScroll: 1 } },
    ],
});
```

```html
<div class="addon-carousel">
    <div class="card lifting-card mx-3">
        <img decoding="async" class="card-img-top p-3 img-fluid" src="..." alt="..." />
        <div class="card-body">
            <h3 class="h5 card-title">Title</h3>
            <p class="card-text">
                Text
            </p>
        </div>
    </div>
</div>
```

**Comment by Aulig** [comment]
I assume the issue comes from this code in line 599 of slick.js:

```javascript
_.$slider.children().children().children()
  .css({
      'width':(100 / _.options.slidesPerRow) + '%',
      'display': 'inline-block'
  });
```

But I don't quite understand why it would only add the classes in the fork, but not the original slick, as the buildRows function (that contains the above code) seems to be unchanged in the fork.

**Comment by socialnicheguru** [comment]
I am using kenweeler/slick 1.8.1 for a project and would like to use this patch. however it does not apply to kenweeler/slick 1.8.1.

the patch fails because of this. can these minor issues be corrected, @nicolaskopp ?

 more slick/slick.js.rej
--- slick/slick.js
+++ slick/slick.js
@@ -1034,7 +1049,7 @@
             .off('focus.slick blur.slick')
             .on(
                 'focus.slick',
-                '*', 
+                '*',
                 function(event) {
                     var $sf = $(this);
 
@@ -1049,7 +1064,7 @@
                 }
             ).on(
                 'blur.slick',
-                '*', 
+                '*',
                 function(event) {
                     var $sf = $(this);

**Comment by socialnicheguru** [comment]
Posted updated path to Drupal.org as I do not have write permissions in the kenwheeler/slick project. @nicolaskopp this is just the modified patch

https://www.drupal.org/project/slick/issues/3300651#comment-14630877
